docs/denotational-design.md

```diff
@@ -1,214 +1,225 @@
+μ findInHistories a histories =
+  filterMaybe (findInHistory a) histories
+
+-- Find a particular artefact is in a history.
+findInHistory :: Eq a => a -> History a -> Maybe a
+μ findInHistory a history = find (== a) (μ history)
+
+-- A special Label that guarantees a starting point, i.e. ~
+root :: Label
+
+-- Get the difference between two directory views.
+diff :: Directory -> Directory -> Diff
+
+-- List the current file or directories in a given Directory view.
+listDirectory :: Directory -> NonEmpty (Label, SystemType)
+μ listDirectory directory = map f $ snd (μ directory)
+  where
+    f = \case
+      Left dir -> (fst dir, IsDirectory)
+      Right file -> (fst file, IsFile)
+
+fileName :: File -> Label
+μ fileName file = fst (μ file)
 
--- Getting the [Directory] from CommitHistory should be equivalent to building
--- the [File] and getting their Directory
-files_and_directories :: Property
-files_and_directories = ∀ history. getHistoryDirectories history ≡ map fileDirectory (getFiles (getCommits history))
+findFile :: NonEmpty Label -> Directory -> Maybe File
+μ findFile (label :| labels) directory =
+  let (label, artefacts) = (μ directory)
+  if label == label' then go labels artefacts else Nothing
+  where
+    findFileWithLabel :: Foldable f => Label -> f (Either Directory File) -> Maybe File
+    findFileWithLabel label = find (\artefact -> case artefact of
+      Left _     -> False
+      Right file -> fileLabel == label)
+
+    go :: [Label] -> [Either Directory File] -> Just File
+    go [] _ = Nothing
+    go [label] directories = findMaybe (fileWithLabel label) directories
+    go (label:labels) directories = go labels $ find ((label ==) . fst) onlyDirectories directories
+
+onlyDirectories :: Foldable f => f (Either Directory File) -> [Directory]
+onlyDirectories = filter isLeft . toList
+
+getSubDirectories :: Directory -> [Directory]
+μ getSubDirectories directory = foldMap f $ snd (μ directory)
+  where
+    f :: Either Directory File -> [Directory]
+    f = either pure []
+
+-- Definition elided
+findDirectory :: NonEmpty Label -> Directory -> Maybe Directory
+
+-- Definition elided
+fuzzyFind :: Label -> [Directory]
+
+-- A Git Snapshot is grabbing the HEAD commit of your History
+-- and turning it into a Directory
+gitSnapshot :: Snapshot [Commit]
+gitSnapshot = getDirectoryPtr . head
+
+-- Opaque and defined by the backend
+getDirectoryPtr :: Commit -> Directory
+
+-- A Pijul history is semantically applying the patches in a
+-- topological order and achieving the Directory view.
+pijulHistory :: History Patch
+pijulHistory = foldl pijulMagic mempty
+
+-- Opaque and defined by the backend
+pijulMagic :: Patch -> Directory -> Directory
+
+-- Get the current History we are working with.
+getHistory :: Browser a (History a)
+getHistory = get
+
+-- Get the current Directory in the Browser
+getDirectory :: Browser a Directory
+getDirectory = do
+  hist <- get
+  fromHistory <- ask
+  pure $ fromHistory hist
```

Comment:
```haskell
ask <$> get
```

:eyes:

Reply:
I think I'll keep the version we have for clarity sake :sweat_smile: But maybe change `fromHistory` to `applySnapshot`?